pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,539 @@
+			_, err := r.client.NetworkingV1alpha1().IPAddresses().Update(context.Background(), newIPAddress(ip.String(), family, svc), metav1.UpdateOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+
+		if !verifyIPAddressLabels(ipaddress) {
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s has wrong labels %v; repairing", family, ip, svc.Name, svc.Namespace, ipaddress.Labels))
+			_, err := r.client.NetworkingV1alpha1().IPAddresses().Update(context.Background(), newIPAddress(ip.String(), family, svc), metav1.UpdateOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *RepairIPAddress) ipWorker() {
+	for r.processNextWorkIp() {
+	}
+}
+
+func (r *RepairIPAddress) processNextWorkIp() bool {
+	eKey, quit := r.ipQueue.Get()
+	if quit {
+		return false
+	}
+	defer r.ipQueue.Done(eKey)
+
+	err := r.syncIPAddress(eKey.(string))
+	r.handleIpErr(err, eKey)
+
+	return true
+}
+
+func (r *RepairIPAddress) handleIpErr(err error, key interface{}) {
+	if err == nil {
+		r.ipQueue.Forget(key)
+		return
+	}
+
+	if r.ipQueue.NumRequeues(key) < maxRetries {
+		klog.V(2).InfoS("Error syncing Service, retrying", "service", key, "err", err)
+		r.ipQueue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping Service %q out of the queue: %v", key, err)
+	r.ipQueue.Forget(key)
+	runtime.HandleError(err)
+}
+
+// syncIPAddress verify that the IPAddress that are owned by the ipallocator controller reference an existing Service
+// to avoid leaking IPAddresses. IPAddresses that are owned by other controllers are not processed to avoid hotloops.
+// IPAddress that reference Services and are part of the ClusterIP are validated in the syncService loop.
+func (r *RepairIPAddress) syncIPAddress(key string) error {
+	ipAddress, err := r.ipAddressLister.Get(key)
+	if err != nil {
+		// nothing to do
+		return nil
+	}
+
+	// not mananged by this controller
+	if !managedByController(ipAddress) {
+		return nil
+	}
+
+	// does not reference a Service but created by the service allocator, something else have changed it, delete it
+	if ipAddress.Spec.ParentRef.Group != "" ||
+		ipAddress.Spec.ParentRef.Resource != "services" {
+		runtime.HandleError(fmt.Errorf("IPAddress %s appears to have been modified, not referencing a Service %v: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef))
+		r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "IPAddressNotAllocated", "IPAddressAllocation", "IPAddress %s appears to have been modified, not referencing a Service %v: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef)
+		err := r.client.NetworkingV1alpha1().IPAddresses().Delete(context.Background(), ipAddress.Name, metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
```

Comment:
is this an error really? what if another thing is updating it, say a delete service operation is executing side by side?

Reply:
the errors are retried, so it is safe, it will go again through the sync loop and reconcile correctly